powerauth-nextstep/src/main/java/io/getlime/security/powerauth/app/nextstep/service/LdapVerifierService.java

```diff
@@ -0,0 +1,180 @@
+                    .searchScope(SearchScope.SUBTREE)
+                    .filter(nextStepServerConfiguration.getUserSearchFilter(), credential.getUser().getUserId());
+            // Spring LDAP will do: search -> resolve DN -> attempt bind with that DN+password
+            ldapTemplate.authenticate(query, credentialValue);
+            return AuthenticationResult.SUCCEEDED;
+        } catch (EmptyResultDataAccessException | NamingException e) {
+            logger.warn("action: verifyCredential, state: failed, reason: {}", e.getMessage(), e);
+            return AuthenticationResult.FAILED;
+        }
+    }
+
+    private static Boolean getBoolAttrSafe(Attributes attrs, String name)  {
+        try {
+            Attribute a = attrs.get(name);
+            if (a == null) {
+                return null;
+            }
+            return Boolean.parseBoolean(a.get().toString());
+        } catch (javax.naming.NamingException e) {
+            return null;
+        }
+    }
+
+    private static Integer getIntAttrSafe(Attributes attrs, String name)  {
+        try {
+            Attribute a = attrs.get(name);
+            if (a == null) {
+                return null;
+            }
+            return Integer.valueOf(a.get().toString());
+        } catch (javax.naming.NamingException | NumberFormatException e) {
+            return null;
+        }
+    }
+
+    private static String getStrAttrSafe(Attributes attrs, String name)  {
+        try {
+            Attribute a = attrs.get(name);
+            if (a == null) {
+                return null;
+            }
+            return a.get().toString();
+        } catch (javax.naming.NamingException e) {
+            return null;
+        }
+    }
+
+    /**
+     * Get attempt counter from LDAP
+     *
+     * @param credential credentials to get attributes for - credentials can have set up special base search to override server configuration.
+     * @return number of attempts.
+     */
+    public ExternalCredentialDetail readCredentialExternalStatus(CredentialEntity credential) throws InvalidRequestException {
+        final String userSearchBase = resolveUserSearchBase(credential);
+
+        if (userSearchBase == null) {
+            logger.error("action: readCredentialExternalStatus, state: failed, reason: {}", CONFIG_ERROR);
+            throw new InvalidRequestException(CONFIG_ERROR);
+        }
+        var query = LdapQueryBuilder.query()
+                .base(userSearchBase)
+                .attributes(PASSWORD_RETRY_COUNT_ATTRIBUTE, ACCOUNT_LOCK_ATTRIBUTE, PASSWORD_EXPIRATION_TIME)
+                .filter(nextStepServerConfiguration.getUserSearchFilter(), credential.getUser().getUserId());
+
+        return ldapTemplate.search(query, (AttributesMapper<ExternalCredentialDetail>) attrs -> {
+            ExternalCredentialDetail externalCredentialDetail = new ExternalCredentialDetail();
+
+            final Boolean blocked = getBoolAttrSafe(attrs, ACCOUNT_LOCK_ATTRIBUTE);
```

Comment:
Do we really need three-state logic? Couldn't we declare the credential ACTIVE when the attribute is missing?

Reply:
That we definitely should not.We simply do not know the state if the state is not shared.